src/stochastic_swap.rs

```diff
@@ -0,0 +1,257 @@
+use crate::nlayout::NLayout;
+
+#[inline]
+fn compute_cost(
+    dist: &ArrayView2<f64>,
+    layout: &NLayout,
+    gates: &[usize],
+    num_gates: usize,
+) -> f64 {
+    (0..num_gates)
+        .map(|kk| {
+            let ii = layout.logic_to_phys[gates[2 * kk]];
+            let jj = layout.logic_to_phys[gates[2 * kk + 1]];
+            dist[[ii, jj]]
+        })
+        .fold(0.0, |a, b| a + b)
+}
+
+/// Computes the symmetric random scaling (perturbation) matrix,
+/// and places the values in the 'scale' array.
+///
+/// Args:
+///     scale (ndarray): An array of doubles where the values are to be stored.
+///     cdist2 (ndarray): Array representing the coupling map distance squared.
+///     rand (double *): Array of rands of length num_qubits*(num_qubits+1)//2.
+///     num_qubits (int): Number of physical qubits.
+#[inline]
+fn compute_random_scaling(
+    scale: &mut Array2<f64>,
+    cdist2: &ArrayView2<f64>,
+    rand: &[f64],
+    num_qubits: usize,
+) {
+    let mut idx: usize = 0;
+    for ii in 0..num_qubits {
+        for jj in 0..ii {
+            scale[[ii, jj]] = rand[idx] * cdist2[[ii, jj]];
+            scale[[jj, ii]] = scale[[ii, jj]];
+            idx += 1
+        }
+    }
+}
+
+fn swap_trial(
+    num_qubits: usize,
+    int_layout: &NLayout,
+    int_qubit_subset: &[usize],
+    gates: &[usize],
+    cdist: ArrayView2<f64>,
+    cdist2: ArrayView2<f64>,
+    edges: &[usize],
+    seed: Option<u64>,
+    trial_num: u64,
+    locked_best_possible: &RwLock<&mut Option<(f64, EdgeCollection, NLayout, usize)>>,
+) -> Option<(f64, EdgeCollection, NLayout, usize)> {
+    {
+        let best_possible = locked_best_possible.read().unwrap();
+        if best_possible.is_some() {
+            return None;
+        }
+    }
+    let mut opt_edges: EdgeCollection = EdgeCollection::new();
+    let mut new_layout = int_layout.clone();
+    let mut trial_layout = int_layout.clone();
+    let mut optimal_layout = int_layout.clone();
+
+    let num_gates: usize = gates.len() / 2;
+    let num_edges: usize = edges.len() / 2;
+
+    let mut need_copy;
+    let mut cost_reduced;
+    let mut depth_step: usize = 1;
+    let depth_max: usize = 2 * num_qubits + 1;
+    let mut min_cost: f64;
+    let mut new_cost: f64;
+    let mut dist: f64;
+
+    let mut optimal_start: usize = std::usize::MAX;
+    let mut optimal_end: usize = std::usize::MAX;
+    let mut optimal_start_qubit = std::usize::MAX;
+    let mut optimal_end_qubit = std::usize::MAX;
+
+    let mut scale = Array2::zeros((num_qubits, num_qubits));
+
+    let distribution = Normal::new(0.0, 1.0 / num_qubits as f64).unwrap();
+    let mut rng: Pcg64Mcg = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed + trial_num),
+        None => Pcg64Mcg::from_entropy(),
+    };
+    let rand: Vec<f64> = distribution
+        .sample_iter(&mut rng)
+        .take(num_qubits * (num_qubits + 1) / 2)
+        .map(|i| i + 1.0)
+        .collect();
```

Comment:
If you're mapping `|i| i + 1.0`, can't you just set your distribution to have a mean of `1.0` from the start and skip that step?

Reply:
Yeah, that's a fair point. For this first wip pass I very clearly just mechanically ported what was in cython to rust and then made it multithreaded with [rayon](https://github.com/rayon-rs/rayon). This was basically how I read the cython code as working, but I agree we should just change the mean on the distribution and this will be more efficient.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7658/commits/2f8de0190387f9fd6ee1b3765b9bc228a9f0d54f